// eslint-plugin-boundaries/dist/Settings/Settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformLegacyTypes = transformLegacyTypes;
exports.getElementsTypeNames = getElementsTypeNames;
exports.getRootPath = getRootPath;
const node_path_1 = require("node:path");
const Settings_types_1 = require("../Shared/Settings.types");
const Helpers_1 = require("./Helpers");
const { VALID_MODES, ROOT_PATH, ENV_ROOT_PATH } = Settings_types_1.SETTINGS;
/**
 * Converts legacy string element descriptors into object descriptors.
 *
 * @param typesFromSettings - Raw `boundaries/elements` setting value.
 * @returns Normalized element descriptors compatible with current matcher API.
 */
function transformLegacyTypes(typesFromSettings) {
    const types = typesFromSettings || [];
    return types.map((type) => {
        // backward compatibility with v1
        if ((0, Helpers_1.isLegacyType)(type)) {
            return {
                type: type,
                match: VALID_MODES[0],
                pattern: `${type}/*`,
                capture: ["elementName"],
            };
        }
        // default options
        return {
            match: VALID_MODES[0],
            ...type,
        };
    });
}
/**
 * Extracts element type names from normalized element descriptors.
 *
 * @param descriptors - Normalized descriptors from settings.
 * @returns List of defined element type names.
 */
function getElementsTypeNames(descriptors) {
    return descriptors.map((element) => element.type).filter(Boolean);
}
/**
 * Resolves the effective root path used by boundaries matchers.
 *
 * @param settings - Validated plugin settings.
 * @returns Absolute root path used for dependency and element resolution.
 */
function getRootPath(settings) {
    const rootPathUserSetting = process.env[ENV_ROOT_PATH] || settings[ROOT_PATH];
    if (rootPathUserSetting) {
        return (0, node_path_1.isAbsolute)(rootPathUserSetting)
            ? rootPathUserSetting
            : (0, node_path_1.resolve)(process.cwd(), rootPathUserSetting);
    }
    return process.cwd();
}
