// eslint-plugin-boundaries/dist/Rules/External.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const elements_1 = require("@boundaries/elements");
const Settings_1 = require("../Settings");
const Shared_1 = require("../Shared");
const Dependencies_1 = require("./Dependencies");
const Support_1 = require("./Support");
const { RULE_EXTERNAL } = Shared_1.SETTINGS;
/**
 * Type guard for external selectors using tuple syntax with options.
 *
 * @param selector - External library selector from rule options.
 * @returns `true` when selector is `[module, options]`.
 */
function isExternalLibrarySelectorWithOptions(selector) {
    return ((0, Shared_1.isArray)(selector) &&
        selector.length === 2 &&
        (0, Shared_1.isString)(selector[0]) &&
        (0, Shared_1.isObject)(selector[1]));
}
/**
 * Builds a dependency selector from a legacy external selector using tuple syntax with options.
 * @param selector The external library selector in legacy format with options.
 * @returns The corresponding dependency selector compatible with `dependencies` rule evaluator.
 */
function buildSelectorFromLegacySelectorWithOptions(selector) {
    const moduleSelector = selector[0];
    const selectorOptions = selector[1];
    const hasPathSelector = !(0, Shared_1.isNullish)(selectorOptions.path);
    return {
        to: {
            origin: [elements_1.ELEMENT_ORIGINS_MAP.EXTERNAL, elements_1.ELEMENT_ORIGINS_MAP.CORE],
            ...(hasPathSelector ? { internalPath: selectorOptions.path } : {}),
        },
        dependency: {
            module: moduleSelector,
            ...(selectorOptions.specifiers
                ? {
                    specifiers: selectorOptions.specifiers,
                }
                : {}),
        },
    };
}
/**
 * Transforms legacy external selectors into dependency selectors.
 *
 * @param selectors - External selector(s) from legacy rule format.
 * @returns Dependency selector(s) compatible with `dependencies` rule evaluator.
 */
function modifySelectors(selectors) {
    const originsToMatch = [
        elements_1.ELEMENT_ORIGINS_MAP.EXTERNAL,
        elements_1.ELEMENT_ORIGINS_MAP.CORE,
    ];
    if (isExternalLibrarySelectorWithOptions(selectors)) {
        return buildSelectorFromLegacySelectorWithOptions(selectors);
    }
    if ((0, Shared_1.isString)(selectors)) {
        return {
            to: {
                origin: originsToMatch,
            },
            dependency: {
                module: selectors,
            },
        };
    }
    return selectors.map((selector) => {
        if (isExternalLibrarySelectorWithOptions(selector)) {
            return buildSelectorFromLegacySelectorWithOptions(selector);
        }
        return {
            to: { origin: originsToMatch },
            dependency: {
                module: selector,
            },
        };
    });
}
/**
 * Converts `external` legacy rules to `dependencies` rule shape.
 *
 * @param rules - External rules as configured by the user.
 * @returns Equivalent dependencies rules consumed by shared evaluator.
 */
function transformToDependenciesRules(rules) {
    return rules.map((rule) => ({
        from: rule.from,
        allow: rule.allow ? modifySelectors(rule.allow) : undefined,
        disallow: rule.disallow ? modifySelectors(rule.disallow) : undefined,
        importKind: rule.importKind,
        message: rule.message,
    }));
}
exports.default = (0, Support_1.dependencyRule)({
    ruleName: RULE_EXTERNAL,
    description: `Check dependencies to external and core libraries`,
    schema: (0, Settings_1.rulesOptionsSchema)({
        isLegacy: true,
        targetMatcherOptions: {
            type: "object",
            properties: {
                specifiers: {
                    type: "array",
                    items: {
                        type: "string",
                    },
                },
                path: {
                    anyOf: [
                        {
                            type: "string",
                        },
                        {
                            type: "array",
                            items: {
                                type: "string",
                            },
                        },
                    ],
                },
            },
            additionalProperties: false,
        },
    }),
}, function ({ dependency, node, context, settings, options }) {
    (0, Settings_1.warnMigrationToDependencies)(Shared_1.RULE_NAMES_MAP.EXTERNAL);
    // Validate and warn about legacy selector syntax
    (0, Settings_1.validateAndWarnRuleOptions)(options, Shared_1.RULE_NAMES_MAP.EXTERNAL, "from");
    if ((0, elements_1.isExternalDependencyElement)(dependency.to) ||
        (0, elements_1.isCoreDependencyElement)(dependency.to)) {
        const rules = transformToDependenciesRules(options?.rules ?? []);
        (0, Dependencies_1.evaluateRulesAndReport)({
            rules,
            settings,
            context,
            node,
            options,
            dependency,
        });
    }
}, {
    validateRules: { onlyMainKey: true },
});
