// eslint-plugin-boundaries/dist/Rules/NoUnknownFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Elements_1 = require("../Elements");
const Settings_1 = require("../Settings");
const Shared_1 = require("../Shared");
const Support_1 = require("./Support");
const { RULE_NO_UNKNOWN_FILES } = Shared_1.SETTINGS;
const noUnknownFilesRule = {
    ...(0, Support_1.meta)({
        ruleName: RULE_NO_UNKNOWN_FILES,
        schema: [],
        description: `Prevent creating files not recognized by any element patterns`,
    }),
    create: function (context) {
        const settings = (0, Settings_1.getSettings)(context);
        const file = (0, Elements_1.elementDescription)(context.filename, settings);
        if (file.isIgnored || !file.isUnknown) {
            return {};
        }
        return {
            Program: (node) => {
                context.report({
                    message: `File does not match any element pattern`,
                    node: node,
                });
            },
        };
    },
};
exports.default = noUnknownFilesRule;
